scipy/signal/spectral.py

```diff
@@ -952,3 +951,40 @@ def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft):
     result = fftpack.fft(result, n=nfft)
 
     return result
+
+def _triage_segments(window, nperseg):
+    """
+    private function that handles defaults.
+    called by _spectral_helper, and also by spectrogram because it requires a
+    value for nperseg when setting noverlap to the default of nperseg // 8
+    If window is specified by a string or tuple and nperseg is not specified, 
+    _triage_segments sets nperseg to the default of 256 and the window is that
+    size. If instead the window is array_like and nperseg is not specified, then
+    nperseg is set to the length of the window.
+    
+    Returns
+    -------
+    window : ndarray
+        Array of FFT data
+    
+    nperseg : int, optional
+        Length of each segment.  Defaults to 256.
+    """
+    
+    #parse window; if array like, then set nperseg = win.shape
+    if isinstance(window, string_types) or isinstance(window, tuple):
+        # if nperseg not specified
+        if nperseg is None:  
+            nperseg = 256  # then change to default
+        win = get_window(window, nperseg)
+    else:
+        win = np.asarray(window)
+        if len(win.shape) != 1:
+            raise ValueError('window must be 1-D')
+        if nperseg is None:
+            nperseg = win.shape[0]
+        elif nperseg is not None:
+            if nperseg != win.shape[0]:
+                raise ValueError("Value specified for nperseg is different from"
+                                 " length of ndarray provided as window.")
```

Comment:
The window does not, strictly speaking, need to be provided as an `ndarrary`. It's probably clearer to say something like "window must have length of nperseg."